p3387.tex

```diff
@@ -256,7 +263,8 @@ \subsection{Treat coroutine-ness as an implementation detail}
+Therefore, making a parameter that is declared \tcode{const} on the declaration non-\tcode{const} in the definition is indeed an implementation detail that is observable in \tcode{post}, but that is already the case with the specification of \cite{P2900R8} today,\footnote{Note that with \cite{P2900R8} as specified today, if the compiler sees any declaration of a function \tcode{f}, which includes any definition of \tcode{f}, where a given non-reference parameter is not \tcode{const}, in any translation unit, it already must diagnose any \tcode{post} on \tcode{f} that odr-uses that parameter in its predicate as ill-formed, even if that \tcode{post} is on another declaration of \tcode{f} in another translation unit and that declaration declares the parameter as \tcode{const}. ???Solution 3 therefore does not add any additional implementation burden.} and can happen irrespective of whether the function is a coroutine or not. Solution 3 therefore does not expose the coroutine-ness as such to the caller; it merely acknowledges that making a function a coroutine is one of the possible implementations that make a parameter non-\tcode{const}. The litmus test here is whether Solution 3 would give us a way to detect at the call site of a function whether that function is a coroutine. Solutions 1, 2, 4, and 5c would do this; Solution 3 would not. We therefore conclude that Solution 3 does not violate the principle that coroutine-ness is an implementation detail.
+%IDS I do not completely agree with the observation that there's no additional implementation burden - one of the trickier parts (and another questionable design choice) of coroutines is that they are lazily discovered by use of co_zzzz keywords - so that when the function signature is processed (including the contracts) we do not yet know that we have a coroutine.  Currently, we would emit relevant diagnostics as we encounter ill-formed parm-use when parsing contract checks, but we cannot know that we have an ill-formed case for a coroutine until some time later - so we need an additional layer of diagnostic work there.  [FWIW; if it were trivial to do I'd have already added it to Nina's prototype].
```

Comment:
I think my point here was that the implementation challenge you describe already exists today. For example, if a function is a coroutine, it cannot have a non-copyable non-moveable type such as std::atomic as a parameter; but you only discover this when you parse the body and find one of the co_xxx keywords. It seems like it's exactly the same challenge here?

Reply:
I do not expect it to be a challenge - as such - but it should be understood that it's not covered by the existing diagnostics for contract check validity (we need to wire something up to do that later).  The situation you mention does not actually require any special action for a coroutine - the action of trying to make the copy will trigger diagnostics provided by the core compiler (at least for GCC).

Reply:
Okay, should I just remove "Solution 3 therefore does not add any additional implementation burden" then? Or is there some other edit that you would prefer here? Thanks

Reply:
I'd be OK with "does not add significant additional" .. IDK if any clang folks are reading this...

Reply:
OK, what about "does not seem to add significant additional"

Reply:
works for me.